Propose ROS 2 launch front-end design hints. (#208)
ros2/design: articles/160_roslaunch_frontend.md

permalink: articles/roslaunch.html
abstract:
  The launch system in ROS 2 aims to support extension of static descriptions, so as to easily allow both exposing new features of the underlying implementation, which may or may not be extensible itself, and introducing new markup languages. This document discusses several approaches for implementations to follow.
author: '[Michel Hidalgo](https://github.com/hidmic) [William Woodall](https://github.com/wjwwood)'
published: true
---

- This will become a table of contents (this text will be scraped).
{:toc}

# {{ page.title }}

<div class="abstract" markdown="1">
{{ page.abstract }}
</div>

Authors: {{ page.author }}

## Context

Static launch descriptions are an integral part to ROS 2 launch system, and the natural path to transition from predominant ROS 1 `roslaunch` XML description.
This document describes parsing and integration approaches of different front ends i.e. different markup languages, with a focus on extensibility and scalability.

## Proposed approaches

In the following, different approaches to the solution are described.
This list is by no means exhaustive.

> It's worth to note some commonalities between them:

Reviewer: ```suggestion
It's worth noting some things they all have in common:
```

Author: Suggestion accepted (but I'm skipping the Github suggestion to avoid DCO issues).

  parsers to lookup.
  How that registry is populated and provided to the parser may vary.
  For instance, in Python class decorators may populate a global dict or even its import
  mechanism may be used if suitable, while in C++ convenience macros may expand into
  demangled registration hooks that can later be looked up by a dynamic linker (assuming
  launch entities libraries are shared libraries).

### Forward Description Mapping (FDM)

In FDM, the parser relies on a schema and well-known rules to map an static description (markup) to implementation specific instances (objects).

#### Description Markup

Some description samples in different markup languages are provided below:

*XML*
```xml
<launch version="x.1.0">
  <action name="some-action" type="actions.ExecuteProcess">
    <arg name="cmd" value="/bin/ls"/>
  </action>
  <on_event type="events.ProcessExit" target="some-action">
    <action type="actions.LogInfo">
      <arg name="message" value="I'm done"/>
    </action>
  </on_event>
</launch>
```

*YAML*
```yml
launch:
  version: x.1.0
  actions:
    - type: actions.ExecuteProcess
      name: some-action
      args:
        cmd: /bin/ls

>   handlers:

Reviewer: ```suggestion
  event_handlers:
```

Author: Suggestion accepted (but I'm skipping the Github suggestion to avoid DCO issues).

*-* Statically typed launch system implementations may require variant objects to deal with actions.

*-* Static descriptions are geared towards easing parsing, making them more uniform like a serialization format but also less user friendly.

*-* Care must be exercised to avoid coupling static descriptions with a given implementation.

### Forward Description Mapping plus Markup Sugars (FDM+)

A variation on FDM that allows launch entities to supply markup language specific hooks to do their own parsing.

#### Description Markup

Some description samples in different markup languages are provided below:

*XML*
```xml
<launch version="x.1.0">
  <process name="some-action" cmd="/bin/ls">
     <on_exit>
        <log message="I'm done"/>
     </on_exit>
  </process>
</launch>
```

*YAML*
```yml
launch:
  version: x.1.0
  entities:
    - process:
        cmd: /bin/ls
        on_exit:
          - log: "I'm done"
```

#### Parsing Hooks

Some code samples in different programming languages are provided below:

*Python*
```python
class SomeAction(LaunchDescriptionEntity):

   @launch.parse_from('xml')

>    def parse(cls, element):

Reviewer: nitpick: I know this is a just a theoretical example, but it might be good (even in the examples) to separate the markup helpers (markup sugar or w/e you want to call them) from the actual action.

Something like:

```python
@launch_xml.handle_tag('process')
def process_tag_helper(xml_element):
    ...
    return launch.actions.ExecuteProcess(...)
```

Author: Yeah, I see where you're going and it makes sense. Changed.

       ...
```

*C++*
```c++
class SomeAction : public LaunchDescriptionEntity {
public:
  static std::unique_ptr<SomeAction> parse_from_xml(const XmlElement& e) {
     ...
  }
};

> LAUNCH_XML_PARSING_HOOK("some-action", SomeEntity::parse_from_xml);

Reviewer: For this one it's be like:

```c++
std::unique_ptr<launch_cpp::actions::ExecuteProcess>
process_tag_helper(const XmlElement & xml_element)
{
  // ...
  return std::make_unique<launch_cpp::actions::ExecuteProcess>(/* ... */);
}

LAUNCH_CPP_XML_HANDLE_TAG('process', process_tag_helper);
```

Reviewer: Though having written both examples now, I wonder if this doesn't overlap with some proposal further down in the article...

Author: Same as above.

In the following, different approaches to parsing launch descriptions are described.
This list is by no means exhaustive.

It's worth noting some things they all have in common:

- All of them attempt to solve the problem in a general and extensible way to help the system scale with its community.

- All of them require a way to establish associations between entities, solved using unique reference id (usually, a human-readable name but that's not a requisite).

- All of them associate a markup language with a given substitution syntax.

- All of them supply a general, interpolating substitution to deal with embedded substitutions e.g. "rooted/$(subst ...)".

- All of them need some form of instantiation and/or parsing procedure registry for parsers to lookup.
  How that registry is populated and provided to the parser may vary.
  For instance, in Python class decorators may populate a global dict or even its import mechanism may be used if suitable, while in C++ convenience macros may expand into demangled registration hooks that can later be looked up by a dynamic linker (assuming launch entities libraries are shared libraries).

### Forward Description Mapping (FDM)

> In FDM, the parser relies on a schema and well-known rules to map an static description (markup) to implementation specific instances (objects).

Reviewer: ```suggestion
In FDM, the parser relies on a schema and well-known rules to map a static description (markup) to implementation specific instances (objects).
```

  <action name="some-action" type="actions.ExecuteProcess">
    <arg name="cmd" value="{% substitutions.FindExecutable name=my-process %}"/>
    <arg name="env">
      <pair name="LD_LIBRARY_PATH"
            value="/opt/dir:{% substitutions.EnvironmentVariable name=LD_LIBRARY_PATH %}"/>
    </arg>
    <arg name="prefix" value="{% substitutions.EnvironmentVariable name=LAUNCH_PREFIX %}"/>
  </action>
  <on_event type="events.ProcessExit" target="some-action">
    <action type="actions.LogInfo">
      <arg name="message" value="I'm done"/>
    </action>
  </on_event>
</launch>
```

*YAML*
```yml
launch:
  version: x.1.0
  actions:
    - type: actions.ExecuteProcess
      name: some-action
      args:
        cmd: $(substitutions.FindExecutable name=my-process)
        env:
          LD_LIBRARY_PATH: "/opt/dir:$(substitutions.EnvironmentVariable name=LD_LIBRARY_PATH)"
        prefix: $(substitutions.EnvironmentVariable name=LAUNCH_PREFIX)
  event_handlers:
    - type: events.ProcessExit
      target: some-action
      actions:
        - type: actions.LogInfo
          args:
            message: "I'm done"
```

#### Advantages & Disadvantages

*+* Straightforward to implement.

> *+* Launch implementations are completely unaware of the static description existence and its parsing process (to the extent that type agnostic instantiation mechanisms are available).

Reviewer: ```suggestion
*+* Launch implementations are completely unaware of the existence of the static description formats and their parsing processes (to the extent that type agnostic instantiation mechanisms are available).
```

*Python*
```python
@launch_markup.xml.handle_tag('process')
def process_tag_helper(xml_element, parser):
    ...
    return launch.actions.ExecuteProcess(...)

@launch_markup.handle_subst('env')
def env_subst_helper(args, parser):
    ...
    return launch.substitutions.EnvironmentVariable(parser.parse(args[0]))
```

*C++*
```c++
std::unique_ptr<launch::actions::ExecuteProcess>
process_tag_helper(const XmlElement & xml_element, const launch_markup::xml::Parser & parser) {
    // ...
    return std::make_unique<launch::actions::ExecuteProcess>(/* ... */);
};

LAUNCH_XML_MARKUP_HANDLE_TAG("process", process_tag_helper);

std::unique_ptr<launch::substitutions::EnvironmentVariable>
env_subst_helper(const std::vector<std::string> & args, const launch_markup::Parser & parser) {
    // ...
    return std::make_unique<launch::substitutions::EnvironmentVariable>(parser.parse(args[0]));
}

LAUNCH_MARKUP_HANDLE_SUBST("env", env_subst_helper);
```

#### Advantanges & Disadvantages

*+* Straightforward to implement.

*-* Launch system implementations are aware of the parsing process, being completely involved with it if sugars are to be provided.

> *+* Allows leveraging the strenghts of each markup language.

Reviewer 2: ```suggestion
*+* Allows leveraging the strengths of each markup language.
```

> *-* Opens the door to big differences in the representation of launch entities across different front end, and even within a given one by allowing the users to introduce multiple custom representations for the same concepts (e.g. a list of numbers).

Reviewer 2: ```suggestion
*-* Opens the door to big differences in the representation of launch entities across different front ends, and even within a given one by allowing the users to introduce multiple custom representations for the same concept (e.g. a list of numbers).
```

*Python*
```python

> @launch.frontend.expose_action('some-action')  # Infers it's an Action

Author: @ivanpauno I think we should remove `# Infers it's an Action`.

class SomeAction(launch.Action):

    @classmethod
    def parse(
        cls,
        entity: launch.frontend.Entity,
        parser: launch.frontend.Parser
    ) -> SomeAction:
        return cls(
            scoped=parser.parse_substitution(entity.get_attr(scoped)), entities=[
                parser.parse_action(child) for child in entity.children
            ]
        )

> @launch.frontend.expose_substitution('some-subst')  # Infers it's a Substitution

Author: @ivanpauno I think we should remove `# Infers it's a Substitution`.

*+* The static description abstraction effectively decouples launch frontends and backends, allowing for completely independent development and full feature availability at zero cost.

*-* No markup language specific sugars are possible.
    REVISIT(hidmic): IMHO explicitly disallowing this is a good thing, it makes for more homogeneus descriptions and avoids proliferation of multiple representation of the same concepts (e.g. a list of strings).

> 

Author: @ivanpauno this still applies.
